Feature: add manual mode to booth tickets

ChrisKowalczyk opened the pull request:
This pull request contains the changes, required for handling manual tickets in GEO clusters, as discussed in [Issue 60](https://github.com/ClusterLabs/booth/issues/60). The changes modify some most basic Booth functionalities related to Raft algorithm, so I would be happy to hear all comments regarding rearranging the changes or adding new modifications.

dmuhamedagic commented:
In general, looks good. I'll do a more thorough review and add
some more specific comments.

How about adding a test case for the feature? Do all current
tests pass (I'm referring to live_test)?

ChrisKowalczyk commented:
All existing tests are okay, as the behavior for non-manual tickets has not been changed. I am going to add the tests for manual mode, but first I wanted to confirm the implementation & behavior by presenting you the changes in this pull request.
Would you like to add those tests to this PR or to create a new one?

dmuhamedagic commented:
On Mon, Aug 28, 2017 at 08:50:02AM +0000, Chris Kowalczyk wrote:
> All existing tests are okay, as the behavior for non-manual tickets has not been changed. I am going to add the tests for manual mode, but first I wanted to confirm the implementation & behavior by presenting you the changes in this pull request.
> Would you like to add those tests to this PR or to create a new one?

Either way is fine, however it is easier for you. Let me know if you need help with it.

dmuhamedagic commented on src/config.c:
```diff
@@ -331,6 +332,16 @@ static int parse_weights(const char *input, int weights[MAX_NODES])
+static ticket_mode_e retrieve_ticket_mode(const char *input)
+{
+	if ((strcmp(input, "manual") == 0) ||
+	    (strcmp(input, "MANUAL") == 0)) {
```
strcasecmp? Actually, I cannot recall if other stuff in the configuration file considers the case at all. Best to try to be consistent with what's already there.

ChrisKowalczyk replied:
sure, will use strcasecmp instead

dmuhamedagic commented on src/config.c:
```diff
@@ -546,6 +557,7 @@ int read_config(const char *path, int type)
 	defaults.timeout       = DEFAULT_TICKET_TIMEOUT;
 	defaults.retries       = DEFAULT_RETRIES;
 	defaults.acquire_after = 0;
+	defaults.mode  = 0;  /* by default, ticket management is automatic */
```
Shouldn't this be TICKET_MODE_AUTO?

ChrisKowalczyk replied:
ok

dmuhamedagic commented on src/manual.c:
```diff
@@ -0,0 +1,64 @@
+	tk_log_debug("starting manual selection");
+	tk_log_debug("selection caused by %s %s",
+				state_to_string(reason),
+				reason == OR_AGAIN ? state_to_string(tk->election_reason) : "" );
```
This is just a debug message, but still: if there are no elections on manual tickets, what can this message disclose?

ChrisKowalczyk replied:
it was more to keep this consistent with election messages. I can remove it if you prefer

dmuhamedagic replied:
Well, if there's nothing new to say... ;-) Also, perhaps it's good that the manual operations have different footprint.

dmuhamedagic commented on src/raft.c:
```diff
@@ -399,10 +399,22 @@ static int process_REVOKE (
-				"but it is not granted there (ignoring)",
-				site_string(sender));
-		return -1;
+		if (tk->mode != TICKET_MODE_MANUAL) {
```
I guess that this test would be repeating often, maybe to add a macro is_manual(tk), it would help with readability.

ChrisKowalczyk replied:
ok

dmuhamedagic commented on src/ticket.c:
```diff
@@ -182,15 +183,27 @@ void save_committed_tkt(struct ticket_config *tk)
+			save_committed_tkt(tk);
+			reset_ticket(tk);
+			ticket_write(tk);	
+			log_error("external test failed on the specified machine, cannot acquire a manual ticket");
```
We should probably make external tests an invalid configuration for manual tickets.

ChrisKowalczyk replied:
do you mean, to disable external tests for manual tickets completely? hm, I am not sure, perhaps this feature may be useful. We could assure that a ticket cannot be granted somewhere, event if someone make a mistake and try to grant it there..

dmuhamedagic replied:
Yes. Well, I'm not sure either. The question is one of balance between simplicity of use and trust, i.e. if the user _wants_ to grant the ticket then let them do that no matter what. Perhaps you can discuss that further also with Yan.

gao-yan replied:
I'd prefer to honor before-acquire-handler also for manual tickets if it's configured, which means the user would still like to double-check with that. Of course, users should still be able to bypass that by supplying "-F" option, or probably "-FF" if there's already another condition against granting that needs to be bypassed ;-)

ChrisKowalczyk replied:
Wouldn't we want to keep this behavior consistent with automatic tickets?
__Automatic tickets__
* when we try to grant a ticket to a particular site and this fails, Booth will start a new election process and try to grant a ticket somewhere else. It may possibly fail altogether if no site can pass external tests.
* we cannot force the ticket to be granted to the site which we prefer. there is no flag that allows us to skip external tests.
__Manual tickets__
* wouldn't it be better to not be able to grant a ticket to the preferred site as well?
* we don't have automatic election, so after failure the user can do the same as Booth for automatic tickets - try to grant it somewhere else. But also without certainty that there is any site that would accept the ticket.

I agree we could have a flag for forcing granting a ticket (skipping external tests) but perhaps we could apply it on a higher level, for both manual and automatic tickets?

gao-yan replied:
Probably yes. Since there could be manual interference for automatic tickets...

dmuhamedagic commented on src/ticket.c:
```diff
@@ -299,7 +312,12 @@ int acquire_ticket(struct ticket_config *tk, cmd_reason_t reason)
 
-	rv = new_election(tk, local, 1, reason);
+	if (tk->mode == TICKET_MODE_MANUAL) {
+		rv = manual_selection(tk, local, 1, reason);
```
The reason can only be one, right?

ChrisKowalczyk replied:
I was thinking about two cases: manual grant or reacquiring a ticket that used to belong to this site but it has been restarted.

dmuhamedagic replied:
I see. But isn't it that even on reacquire it's still the manual ticket, so the rules for manual tickets apply.

dmuhamedagic commented on src/manual.c:
```diff
@@ -0,0 +1,64 @@
+	// Make sure that election_end field is empty
+	time_reset(&tk->election_end);
+
+	if (is_time_set(&tk->delay_commit) && all_sites_replied(tk)) {
```
Was there anything broadcasted before so that we should expect replies?

ChrisKowalczyk replied:
hm, acks for heartbits? in the unlikely case when ticket was already granted.. I agree, not a very likely or important scenario.. Can remove if you prefer

dmuhamedagic replied:
Can we then check only in that case? Or are the acks important at all?

ChrisKowalczyk replied:
After some thinking, perhaps for manual tickets we should always reset __delay_commit__ (that is, say, to not set __delay_commit__ at all for manual tickets).  After all, we don't care for the responses from other nodes, a manual ticket is granted immediately after the command.

gao-yan replied:
In the case of granting a granted ticket, is it probably supposed to receive certain reply from the previous holder about a successful/failed revocation so that it knows whether it's safe to grant now?

gao-yan replied:
Of course it probably doesn't matter if re-grant is not the combination of revoke&grant ;)

ChrisKowalczyk replied:
Granting a granted ticket doesn't trigger another leader to send any kind of confirmation.
So, I guess we don't have to wait for any messages here (as I mentioned a few comments above). I will remove this IF and always clear tk->delay_commit time

dmuhamedagic commented on src/manual.c:
```diff
@@ -0,0 +1,64 @@
+	tk->ticket_updated = 0;
+
+	return 0;
+}
```
From the implementation viewpoint, this is an exception to normal ticket management, i.e. the ticket is granted synchronously. Maybe that's fine, but maybe it would be beneficial to go through the usual asynchronous path, i.e. ticket_cron().

ChrisKowalczyk replied:
I was thinking about that. However, in case of manual tickets we don't have distinction like for automatic tickets. For example, leader for automatic ticket would expect acks or send heartbeats. A candidate would wait for the votes.
What could apply for a manual ticket, which is being granted? No matter what was its state before, the ticket could only call 'complete selection' function, which would contain the second part of this function.
We could also do something like that: at the beginning of selection set it to 'candidate', wait a while, and set it to 'leader'. Would you prefer something like this?

dmuhamedagic replied:
I really cannot say. We'd need to review what all housekeeping tasks are being done for the normal tickets and then see which of those apply for manual too and finally weigh on what is mor practical. The danger of having two paths is that it could happen that in future both of them may need to be updated. Maybe we just leave it as is for now and review a bit later again, once all other bits are in place. Also, IIRC, Yan was talking about honoring certain delay parameter even for manual grants, but cannot recall which one.

ChrisKowalczyk replied:
For manual tickets we want to still honor __acquire_after__ time, which is not really used for granting tickets. This time only tells Booth when it should check what is happening with the ticket.
I think the the whole election for manual tickets (which I called _manual selection_ in code), do not have to bother with any delays, while granting manual tickets, so we do not have to use all those ticket_cron() mechanism here.
However, if you want, I can add this step, as mentioned in my previous comment. But I rather think we don't need it :)

gao-yan replied:
Not sure how much of the existing logic would be possible/worth to be reused. But of course it'd be beneficial to somehow unify the basic workflow on handling all kinds of tickets  and potentially simply the maintenance of the code.

The case of granting a granted ticket is closer to the existing workflow I guess. But of course, it still depends on how we'd like to handle it. For example, would it be more convenient to use the existing workflow to honor "acquire_after"?  Could the case possible/necessary to either be split into two steps, revoke&grant?

gao-yan replied:
Well, now IIUC, re-grant is not the combination of revoke & grant, right? If so, it's not necessary to honor acquire_after I assume.

dmuhamedagic replied:
I wonder when would re-grant (i.e. grant an already granted ticket) happen. Is there a plausible scenario for that? Given that booth doesn't have a configuration reload ... well, what actually happens on booth start when it sees a manual ticket in the configuration?

ChrisKowalczyk replied:
re-grant shouldn't be very often. I can imagine two scenarios:
1) One site with granted ticket is down. Admin grants this ticket to another site. (it is not possible to revoke the ticket because the site is down).
2) Ticket is granted to a site. And _for some strange reason_ admin grants the same ticket to another site. We are ending up having a ticket granted to two sites.

Regarding reading the configuration.. There is not much difference between manual and automatic tickets during startup. We just set a flag for the ticket, really nothing more.

gao-yan replied:
Sorry for the confusion. It was the scenario of "move a granted ticket to somewhere else" in my mind ;-) But it seems it's not able to do a "move" as a batch operation yet.

dmuhamedagic replied:
Yan, if I recall correctly, I was asked once to implement the "move" command, but argued against it.

gao-yan replied:
Sounds somewhat tricky ;-)

ChrisKowalczyk replied:
On the one hand, it would be helpful to have 'move' command, on the other hand when you think what impact it could have on many underlying resources.. perhaps it is better to not make it easier for the user to move tickets :)

dmuhamedagic replied:
I'd say so too.

dmuhamedagic commented on src/ticket.c:
```diff
@@ -373,14 +402,15 @@ int list_ticket(char **pdata, unsigned int *len)
 	cp = data;
 	foreach_ticket(i, tk) {
-		if (is_time_set(&tk->term_expires)) {
+		if (is_time_set(&tk->term_expires) && (tk->mode != TICKET_MODE_MANUAL)) {
```
In general (i.e. not only here), I'd move the is_manual() predicate to the front of the if statement. That way you guarantee that no code is going to be executed unwanted.

ChrisKowalczyk replied:
sure, good idea!

dmuhamedagic commented on src/ticket.c:
```diff
@@ -434,17 +470,6 @@ void disown_ticket(struct ticket_config *tk)
-
-	return 0;
-}
-
```
This is unused code removed. Normally, you keep the unrelated changes separate.

ChrisKowalczyk replied:
sounds good to me, sorry for the confusion!

dmuhamedagic replied:
It is often tough to restrain oneself when one runs into something that can obviously and immediately be fixed. I do that often myself and then go back reverting the change :)

ChrisKowalczyk commented on docs/boothd.8.txt:
```diff
@@ -516,6 +527,48 @@ configuration separately using 'systemctl':
+MANUAL TICKETS
+-----------------------
+Manual tickets allow users to create and manage tickets which behave
+more predictably than standard Raft tickets. Granting and revoking
```
ok

ChrisKowalczyk commented on docs/boothd.8.txt:
```diff
@@ -516,6 +527,48 @@ configuration separately using 'systemctl':
+manual tickets if fully controlled by the administrator. It is possible
+to define a number of manual and normal tickets in one GEO cluster.
+
+Automatic ticket management provided by Raft algorithm doesn't apply
```
ok

ChrisKowalczyk commented on docs/boothd.8.txt:
```diff
@@ -516,6 +527,48 @@ configuration separately using 'systemctl':
+to manually controlled tickets. In particular, there is no elections,
+automatic failover procedures, and term expiration.
+
+On the other hand, 'booth' controls if a ticket is currently being
```
ok

ChrisKowalczyk commented on docs/boothd.8.txt:
```diff
@@ -516,6 +527,48 @@ configuration separately using 'systemctl':
+
+Also, configuring only manual tickets in a GEO cluster, allows to have
+just two sites in a cluster, without a need of having an arbitrator.
+This happens because there is no automatic elections and no voting
```
ok

ChrisKowalczyk commented on docs/boothd.8.txt:
```diff
@@ -516,6 +527,48 @@ configuration separately using 'systemctl':
+Manual tickets can be granted and revoked by using normal 'grant' and
+'revoke' commands, with the usual flags and parameters. The only
+difference is that specyfiyng '-F' flag during 'grant' command, forced
+a site to become a leader of the specified ticket, irrespective of the
```
ok

ChrisKowalczyk commented:
@dmuhamedagic, thanks for looking at that. I've fixed some of the issues and I have added my comments to the others. I will squeeze the commits later, to make this changes more clean.
Please, have a look at the current version.

dmuhamedagic commented:
Well, if there's nothing new to say... ;-) Also, perhaps it's good that the manual operations have different footprint.

dmuhamedagic commented:
Welcome. A bit busy with some other stuff now, but will try by the end of the week.

dmuhamedagic commented on src/ticket.c:
```diff
@@ -1159,42 +1199,55 @@ void set_ticket_wakeup(struct ticket_config *tk)
+				ticket_next_cron_at(tk, &tv);
+			} else {
+				/* Set next cron some half minute after acquire_after time */
+				ticket_next_cron_in(tk, 30*TIME_RES + tk->acquire_after);
```
There being no failover for manual tickets, is this needed at all? Shouldn't manual tickets just be idle, i.e. all further processing obsoleted?

ChrisKowalczyk replied:
Agreed. As a follower, we don't expect any actions, so we shouldn't plan any cron tasks here.

dmuhamedagic commented on src/ticket.c:
```diff
@@ -955,6 +987,12 @@ static void next_action(struct ticket_config *tk)
+	} else {
+		/* for manual tickets, just send ACKs if needed */
+		if (tk->acks_expected) {
+			handle_resends(tk);
```
Resends are handled, IIRC, only by the side which initiates the operation. So, I wonder if this is necessary for manual tickets.

gao-yan commented on docs/boothd.8.txt:
```diff
@@ -516,6 +527,49 @@ configuration separately using 'systemctl':
+MANUAL TICKETS
+-----------------------
+Manual tickets allow users to create and manage tickets which are
+subsequently not controlled by booth more predictably than standard
```
Somehow to rephrase this? ;-)

gao-yan commented on docs/boothd.8.txt:
```diff
@@ -516,6 +527,49 @@ configuration separately using 'systemctl':
+-----------------------
+Manual tickets allow users to create and manage tickets which are
+subsequently not controlled by booth more predictably than standard
+Raft tickets. Granting and revoking manual tickets if fully controlled
```
s/if/are/ ?

gao-yan commented on src/config.h:
```diff
@@ -57,6 +57,11 @@ typedef enum {
 } grant_type_e;
 
+typedef enum {
+	TICKET_MODE_AUTO = 1,
```
Doesn't pratically matter, but 0 seems good for the first enumerator to me ;-)

ChrisKowalczyk replied:
Well, you would be surprised ;) 

> [...]
> typedef enum {
> 	ATTR_OP_EQ = 1,
> 	ATTR_OP_NE,
> } attr_op_e;
> 
> typedef enum {
> 	GRANT_AUTO = 1,
> 	GRANT_MANUAL,
> } grant_type_e;
> 
> typedef enum {
> 	TICKET_MODE_AUTO = 1,
> 	TICKET_MODE_MANUAL,
> } ticket_mode_e;
> [...]

gao-yan commented on src/main.c:
```diff
@@ -942,6 +942,8 @@ static void print_usage(void)
 	"  -s <site>     Connect/grant to a different site\n"
 	"  -F            Try to grant the ticket immediately\n"
 	"                even if not all sites are reachable\n"
+	"                Only for manual tickets:\n"
```
Would it be less confusing without "Only"? ;-)

ChrisKowalczyk replied:
ok

dmuhamedagic commented:
On Mon, Sep 18, 2017 at 12:55:38AM -0700, Gao,Yan wrote:
> gao-yan commented on this pull request.
> 
> 
> +MANUAL TICKETS
> +-----------------------
> +Manual tickets allow users to create and manage tickets which are
> +subsequently not controlled by booth more predictably than standard
> 
> Somehow to rephrase this? ;-)

The "more predictably ..." part seems to have by accident survived.

ChrisKowalczyk commented:
Indeed, sorry for the confusion. I will change that to something like this:
+Manual tickets allow users to create and manage tickets which are
+subsequently handled by booth without using the Raft algorithm.
+Granting and revoking manual tickets is fully controlled

gao-yan commented on src/raft.c:
```diff
@@ -399,10 +399,22 @@ static int process_REVOKE (
+
+			// Ticket itself is not modified. This means that
+			// the current site will still follow another
+			// leader (or be a leader itself).
```
Looks like a tricky situation. Would it make sense to add a notice/warning about it in here?

ChrisKowalczyk replied:
Sure, I will add a warning and improve the comments.

gao-yan commented on src/ticket.c:
```diff
@@ -314,8 +332,18 @@ int do_grant_ticket(struct ticket_config *tk, int options)
-	if (is_owned(tk))
-		return RLT_OVERGRANT;
+	if (is_owned(tk)) {
+		if (is_manual(tk) && (options & OPT_IMMEDIATE)) {
```
Looks like a re-grant won't revoke the ticket first? If so, it means a revocation is expected to be requested by admin first, right? 

It'd be better if we could somehow figure out the current state of the ticket holder. If the current holder is still online, we probably shouldn't re-grant it anyway even if -F is supplied, i.e. enforced granting is allowed only if the holder is offline?

ChrisKowalczyk replied:
The problem is that Booth generally doesn't know if the nodes are online or not.

I would like to add some idea of 'a site holding a ticket being actually online' by using heartbeats (which are sent for both automatic and manual tickets) but perhaps in a separate pull request, after agreeing with the upstream what exactly we want to have.

Perhaps then, we could leave this flow as it is now and open a new _something_ (issue?) to agree how we would like to track onlineness of Booth sites?

dmuhamedagic replied:
Hmm, what do you mean "doesn't know"? There are heartbeats per ticket and booth establishes the state of all nodes on start and keeps tracking that state etc. Even though there won't be a failover, heartbeats should happen for manual tickets too in order for all parties to be able to display correct information about the ticket.

ChrisKowalczyk replied:
Yes, you are right, heartbeats are still sent for manual tickets.
Currently, even if we don't receive a heartbeat for a long time, the node will not be marked as offline. It will loose the ticket but due to the fact that another node organizes and wins the elections for this particular ticket. I am not aware of any mechanism of marking a node as offline in Booth, but perhaps I am missing something?

dmuhamedagic replied:
No, you're not. I see your point now. A node or, to be more precise, a site cannot be marked as offline. Every ticket has its own elections, heartbeats, and the rest. No tickets means no communication. It was inherited and I often considered changing it, but was always worried about introducing instability. Besides, it would be a very big modification. Probably something to look at for v2.0 :)

ChrisKowalczyk replied:
Yeeah, exactly. Al least something like configuration parameter 'consider a site offline after not receiving heartbeats for XX time'.

dmuhamedagic replied:
Well, booth could mark a member lost if it didn't hear from it in a while providing there are some tickets configured for which communication was expected. That shouldn't be too difficult to implement. Shouldn't ;-)

gao-yan replied:
Been thinking about this... Indeed another option is to consider the per-ticket state based on "timeout"/"expire" parameters. But since this is *manual* ticket,  it probably makes sense to keep the logic simple:

As long as there's any holder or potential holder of the ticket:
* Refuses to grant the ticket and gives a warning "Make sure the ticket has been revoked already!"
* With "-F", it grants the ticket with a warning like "the risk is on your own";-)
?

dmuhamedagic replied:
Agreed.

dmuhamedagic commented:
On Mon, Sep 18, 2017 at 03:04:45AM -0700, Chris Kowalczyk wrote:
> Indeed, sorry for the confusion. I will change that to something like this:
> +Manual tickets allow users to create and manage tickets which are
> +subsequently handled by booth without using the Raft algorithm.

But they are not subsequently handled at all, right? :)

ChrisKowalczyk commented:
Well, Booth will still control them but with limited functionalities. For example we will still track if it is granted etc. But I am open for better description :) I am getting to the point that I could event remove those lines altogether! :)

dmuhamedagic commented:
On Tue, Sep 19, 2017 at 07:38:15AM -0700, Chris Kowalczyk wrote:
> re-grant shouldn't be very often. I can imagine two scenarios:
> 1) One site with granted ticket is down. Admin grants this ticket to another site. (it is not possible to revoke the ticket because the site is down).

Right.

> 2) Ticket is granted to a site. And _for some strange reason_ admin grants the same ticket to another site. We are ending up having a ticket granted to two sites.

OK. This case looks quite the same to the first one, at least from the
point of view of booth.

> Regarding reading the configuration.. There is not much difference between manual and automatic tickets during startup. We just set a flag for the ticket, really nothing more.

Note that on restart the site which was holding the manual ticket
before should do a grant automatically. For instance, that can
happen when booth (the resource) moves to another node on one
cluster or on booth restart. That's what booth already does for
automatic tickets.

However, in the case for manual tickets it will need to check too
if another site has the ticket granted, etc, as the site may be
waking up after being down and the ticket was in the meantime
granted elsewhere. Obviously, in case of split brain, the
resolution can be tricky.

In fact, handling tickets on start is already the most complex case.

dmuhamedagic commented:
Ugh, github cannot always handle properly replies to comments per email. Sorry for the confusion, but I hope that it's clear since I quoted parts of your message.

ChrisKowalczyk commented:
Currently, after the scenario:
1) grant ticket to Site A
2) Site A fails
3) grant ticket to Site B
4) Site A is back online

The ticket will be granted to two sites. This will be Administrator's decision to decide from where it should be revoked. I agree, this is not perfect..

dmuhamedagic commented:
On Tue, Sep 19, 2017 at 03:10:21PM +0000, Chris Kowalczyk wrote:
> Currently, after the scenario:
> 1) grant ticket to Site A
> 2) Site A fails
> 3) grant ticket to Site B
> 4) Site A is back online

Maybe site A can make sure that the other parties haven't been
granted the ticket in the meantime. That is easily possible on
restart. On split brain, which the site should also notice, in
case of overgrant, the sites may negotiate who can keep the
ticket.

ChrisKowalczyk commented:
Sorry for late response.  Currently for automatic tickets, we read *whatever is in CIB* (function *int setup_ticket(...)* in *ticket.c*) and set to our ticket. Then the automatic management and timeout solve most of our problems - if there is a new leader, he will take over everything because our ownership has expired.

For manual tickets, we cannot count on timeout. So we have three possibilities:
* **accept everything from CIB** - then it will be up the administrator to make sure the configuration is correct. This is the currently implemented behavior.
* **do not accept anything from CIB** - if there is another leader, we will follow him. If not, there will be no leader (even if before crash we used to be the leader). Again, the administrator will have to make the changes manually.
* **read CIB but do not trust it completely** - this is something that **dmuhamedagic** suggested above. So, we could on startup set the state from CIB, together with some flag (a new field in *ticket structure*). Then, if we receive any heartbeats from other leader, we may revoke the ticket and start to follow that leader (and clear the flag). The disadvantage is the we will have some time of uncertainty, before the site receives messages from other sites (this time may vary).

Handling split brain opens another bag of worms. We would have two (or more) leaders sending each other heartbeats at the same time. It would be a bit dangerous to choose an algorithm to elect the *best* leader. Perhaps, the admin should decide? After all, the idea behind those manual tickets is to give more *control* over resources lying in multiple locations.

I would argue, that the current behavior (we may end up having two leaders but they do not *harm* themselves) is the most conservative and predictable and gives most manual control over tickets, but I will be open to your opinions. **dmuhamedagic**, **gao-yan**?

gao-yan commented:
It makes lot of sense for boothd to check also the state of the manual tickets in the CIB on startup, since that actually directly determines if resources are supposed to be hosted there or not.  But I'd prefer that boothd makes no decision in regard of that.

If a ticket is shown to be already granted in the CIB when boothd starts up, it gives a warning even if the ticket is not granted anywhere else yet, and asks the administrator to either manually confirm the granting (basically to grant it) or manually revoke it (also from CIB). 

If the ticket is indicated to be granted to another site already, it should give a BIG warning about the whole situation of course. But it'll still be up to the administrator to decide what to do regarding that --  cleanly revoke one of them, and grant or confirm the granting for the other.

dmuhamedagic commented:
On Wed, Oct 25, 2017 at 09:40:42AM +0000, Gao,Yan wrote:
> It makes lot of sense for boothd to check also the state of the manual tickets in the CIB on startup, since that actually directly determines if resources are supposed to be hosted there or not.  But I'd prefer that boothd makes no decision in regard of that.

OK.

> If a ticket is shown to be already granted in the CIB when boothd starts up, it gives a warning even if the ticket is not granted anywhere else yet, and asks the administrator to either manually confirm the granting (basically to grant it) or manually revoke it (also from CIB). 

That's an excellent way forward, but boothd runs in the
background, how can it ask the admin to do something? Maybe it
can show some special status for the ticket in the list (say
"waiting for instructions from human/admin").

> If the ticket is indicated to be granted to another site already, it should give a BIG warning about the whole situation of course. But it'll still be up to the administrator to decide what to do regarding that --  cleanly revoke one of them, and grant or confirm the granting for the other.

Right. I think that this sums it up well. The big question is:

On start, how does boothd know for whom the manual ticket is
meant?

At that time there are two sources of information: other booth
servers and the CIB. Some of these sources may have some
information about the ticket, but not necessarily. In general, it
really cannot know, without consensus, and since we don't
necessarily have consensus, then it would be for the best to let
somebody else (i.e. a human) decide what to do with the ticket.

However, there's a problem of intra-site booth failover. I would
expect that in that case boothd knows if the ticket is granted
and where. In particular because that can happen when nobody's
attending to the cluster.

gao-yan commented:
Agreed. I think such warnings about the situations should be first sent to the logs.  And it makes sense to show them for "booth list".

I'm thinking we could either do it this way:

When boothd starts, the first thing is to "probe" the initial/current ticket state according to CIB, and accept the fact that it might be already unexpectedly "granted" in here, and broadcast the ticket state on this site (siteA) to the others. The other sites will accept the declared ticket state on siteA anyway. And the other sites will broadcast the ticket states on them as well... Eventually all the sites will reach the consensus on the ticket states on every sites, and consequently be aware of if there's any "multi-granted", but won't do anything but warning if so.

In that case, every boothd would probably need to  store and maintain the states of a single ticket on all the sites. It'd not need to differentiate the case of intro-site booth failover I assume.

ChrisKowalczyk commented:
Generally, it sounds fine to me.
> When boothd starts, the first thing is to "probe" the initial/current ticket state according to CIB, and accept the fact that it might be already unexpectedly "granted" in here, and broadcast the ticket state on this site (siteA) to the others. 

There are two possible states: _granted_ or _not-granted_. So, if a site considers itself a leader, it will send heartbeats to all other sites, thus making them aware than its state is _granted_. And the states of all remaining sites, which do not send heartbeats, will be considered as _not-granted_. Based on that, we could add proper warnings to command outputs and logs.

ChrisKowalczyk commented:
@dmuhamedagic, would the approach described in the previous two comments be fine for you? That is, from the backend booth perspective we would have 2 leaders, but I would add more warnings and information to logs/command outputs?

dmuhamedagic commented:
Sorry, I thought that we cleared it all up :) I'll review again
the last comments.

dmuhamedagic commented:
On Wed, Oct 25, 2017 at 02:14:05PM +0000, Gao,Yan wrote:
> Agreed. I think such warnings about the situations should be first sent to the logs.  And it makes sense to show them for "booth list".
> 
> I'm thinking we could either do it this way:
> 
> When boothd starts, the first thing is to "probe" the
> initial/current ticket state according to CIB, and accept the
> fact that it might be already unexpectedly "granted" in here,
> and broadcast the ticket state on this site (siteA) to the
> others. The other sites will accept the declared ticket state
> on siteA anyway. And the other sites will broadcast the ticket
> states on them as well... Eventually all the sites will reach
> the consensus on the ticket states on every sites, and
> consequently be aware of if there's any "multi-granted", but
> won't do anything but warning if so.

booth already does that: first query the CIB, then broadcast
asking for the state of tickets from other sites.

> In that case, every boothd would probably need to  store and
> maintain the states of a single ticket on all the sites.

Again, all participants already do maintain the state of all
tickets. Or did you mean something else here?

> It'd
> not need to differentiate the case of intro-site booth failover
> I assume.

As I mentioned, it'd be expected by the users not to have to
intervene in the case of the booth server moving to another node.
If the booth otherwise after start would wait for the
administrator to grant the ticket, then we have to handle the two
cases differently. The question is only whether it's doable :)
But I suppose that if we (booth) find the ticket already granted
to us and if the timestamp is not too old, say withing a few
minutes, then we can try to become the leader for the ticket.
Does that make sense?

dmuhamedagic commented:
On Thu, Oct 26, 2017 at 08:48:29AM +0000, Chris Kowalczyk wrote:
> Generally, it sounds fine to me.
> > When boothd starts, the first thing is to "probe" the initial/current ticket state according to CIB, and accept the fact that it might be already unexpectedly "granted" in here, and broadcast the ticket state on this site (siteA) to the others. 
> 
> There are two possible states: _granted_ or _not-granted_. So,
> if a site considers itself a leader, it will send heartbeats to
> all other sites, thus making them aware than its state is
> _granted_. And the states of all remaining sites, which do not
> send heartbeats, will be considered as _not-granted_. Based on
> that, we could add proper warnings to command outputs and logs.

Yes. Still, as already discussed, we should try to keep in mind
the user's expectations.

gao-yan commented:
Ah, I don't know the details. But awesome, glad to know :)

And in case of manual ticket, it does't even matter if the timestamp is too old or not I assume. If a ticket is already granted in the CIB,  boothd should probably directly accept the fact that it's the leader and leave the ticket there. Of course, maybe it'll find it's not the only leader. That's when we need to give a warning :)

dmuhamedagic commented:
On Thu, Nov 02, 2017 at 03:33:45PM +0000, Gao,Yan wrote:
> Ah, I don't know the details. But awesome, glad to know :)

Yeah, starting is one of the most complex parts.

> And in case of manual ticket, it does't even matter if the
> timestamp is too old or not I assume. If a ticket is already
> granted in the CIB,  boothd should probably directly accept the
> fact that it's the leader and leave the ticket there. Of
> course, maybe it'll find it's not the only leader. That's when
> we need to give a warning :)

But imagine this case: the site that had the ticket granted goes
AWOL. The company evaluates the situation and grants the ticket
to the other site. All good. Now, after a while, the missing site
comes back.

Let's say the power came back. Then, when booth starts, I think
that it shouldn't become a leader, because the granted ticket is
rather stale. Actually, with a stale granted manual ticket it
should consider that case as if the ticket expired, revoke it
from the CIB, and give appropriate warnings in the log and the
ticket listing such as "stale ticket revoked, please grant
again".

Otherwise, let's say that the network connection gets restored
and booth was not restarted. This case seems to be more
complicated. booth would learn that the ticket is granted to the
other site and both sides are going to try to keep the ticket.
The one with the "older" ticket should give up. Not sure, but I
believe that currently booth updates the ticket's timestamp
_only_ if there's quorum. If so, then the timestamps are going to
be respectively the grant time and the split brain time. Then,
the split brain time is obviously older and that booth should
revoke the ticket.

gao-yan commented:
I can tell it should be handled automatically and carefully for automatic tickets. Given that this is a *manual* ticket,  even if "multi-granted"/"multi-active" violation occurs due to whatever reasons, probably we shouldn't touch anything but give warnings, since it's hard to know what is the cause of the situation or what the administrator would really want. They'll have to make the decision which site should keep the ticket themselves...

dmuhamedagic commented:
I'm not sure I can follow, there are quite a few changes. I understand that there's tracking the number of leaders. Not sure, but have a feeling that it's somehow an overkill. At any rate, still not able to gauge the situation :) BTW, why didn't you stuff some calls into set_leader()?

dmuhamedagic commented:
Test refactoring and introduction of multiple tickets per configuration: why did you introduce passing the ticket and conffile? Both tkt and cnf are global variables. With this change, they are both global and occasionally passed to some functions: that's not really necessary and quite confusing too. You can just set tkt/cnf in this new ticket loop.
BTW, this change is essentially unrelated to manual tickets.

dmuhamedagic commented:
On tests for manual tickets: apart from some very simple tests, I doubt that we can reuse other tests for manual tickets. Maybe I'm wrong, but that's my feeling because the circumstances are quite different, that is there is no more consensus or quorum and the outcome of various tests is probably going to be different. We certainly need new ones crafted especially for manual tickets. Say for split brain, leader restart, and similar, i.e. to see that the other site doesn't perform a failover.

dmuhamedagic commented:
Finally, there are a number of "red space blocks" in git diff. Some indentation is done with spaces, please make your editor use only tabs.

dmuhamedagic commented:
Once we have the spacing and variables in the live_test.sh ready, if we all agree, let's merge this and see how booth tics. At any rate, my browser has problems loading this conversation ;-) Chris, great work!

ChrisKowalczyk commented:
Hello **dmuhamedagic**,

> [...] have a feeling that it's somehow an overkill

Sorry for the confusion, this is my fault, I should have described more the idea behind **sites_where_granted** variable. (I will add the comments). Basically, this implements the idea **gao-yan** mentioned above: sitting at the machine in one site, we don't really know what is the state of the manual ticket that has been granted to other sites until we see clear *revoke* commands. Until that, we cannot really say if the other site is down, or if there is a communication error, etc. So we will notify the user that the situation is not clear in the printouts, until we will know (on our site) that there is surely only one site where the ticket is granted.

> [...] why didn't you stuff some calls into set_leader()

I didn't add this to set_leader because this would make this function quite long (a lot of checks if we set a leader or just set noleader) with a lot of functionalities not really relevant to automatic tickets. If you want, I can change that?

> [...] why did you introduce passing the ticket and conffile?

Current tests work only with one ticket defined in the config file. To be able to tests both, manual and automatic, tickets I added splitting the booth config file into multiple config files and testing them one after another. So, it is possible to define a number of tickets in booth config file, and they will all be tested. Also, different tests will be run for manual and automatic tickets.
What we might also like to have is the possibility of testing of multiple tickets. I haven't add any such testing yet (because it would require a lot of modification of the testsuite) but this was my idea, to make the tests more independent, that is to be able to define some tests which would take *2 tickets* or which would take *only one of two tickets present in the conf file*.
But if you want, I can revert some of those changes?

> [...] I doubt that we can reuse other tests for manual tickets

Agreed, we need the tests specified for manual scenarios. I am going to add them, just wanted to show you my idea how to organize tests before spending too much time on them.

> [...] there are a number of "red space blocks" in git diff

Sorry for those indentations, I will look at that and revert those changes!

dmuhamedagic commented:
On Mon, Nov 20, 2017 at 10:52:04AM +0000, Chris Kowalczyk wrote:
> > [...] have a feeling that it's somehow an overkill
> 
> Sorry for the confusion, this is my fault, I should have described more the idea behind **sites_where_granted** variable. (I will add the comments). Basically, this implements the idea **gao-yan** mentioned above: sitting at the machine in one site, we don't really know what is the state of the manual ticket that has been granted to other sites until we see clear *revoke* commands. Until that, we cannot really say if the other site is down, or if there is a communication error, etc. So we will notify the user that the situation is not clear in the printouts, until we will know (on our site) that there is surely only one site where the ticket is granted.

I'm still not sure I follow. Shouldn't that be handled by booth
as is? I mean, it does communicate with other parties for manual
tickets too. If it is a leader and it hears from another booth
that it is a leader too, it can take some action whatever that
may be (try to resolve the multi-leader situation, set some flag
for the ticket). Or is there something else I'm missing?

>> [...] why didn't you stuff some calls into set_leader()
> 
> I didn't add this to set_leader because this would make this function quite long (a lot of checks if we set a leader or just set noleader) with a lot of functionalities not really relevant to automatic tickets. If you want, I can change that?

Why would it be preferable repeating code instead of putting it
in one place?

> > [...] why did you introduce passing the ticket and conffile?
> 
> Current tests work only with one ticket defined in the config file. To be able to tests both, manual and automatic, tickets I added splitting the booth config file into multiple config files and testing them one after another. So, it is possible to define a number of tickets in booth config file, and they will all be tested. Also, different tests will be run for manual and automatic tickets.

Yes, I saw that, but that still doesn't justify modifying the
interface.

> What we might also like to have is the possibility of testing of multiple tickets. I haven't add any such testing yet (because it would require a lot of modification of the testsuite) but this was my idea, to make the tests more independent, that is to be able to define some tests which would take *2 tickets* or which would take *only one of two tickets present in the conf file*.

Right now I cannot imagine testing multiple tickets
simultaneously. The tickets are independent of each other.

Either way is fine: global variables or passing stuff as an
argument. But it should be consistent, i.e. either one or the
other. In this PR it is half-way between.

Since we already have it as global variables and since there
really is no need to pass it as an argument, why spend time
changing it? Shouldn't we wait until there's somebody comes up
with multiple ticket tests?

ChrisKowalczyk commented:
Ok, I will move back as much as possible to global variables in live_tests and move everything related to sites_where_granted to set_leader()

dmuhamedagic commented:
Great! Many thanks for the contribution. Let's see now what the
testing shows :)

ChrisKowalczyk commented:
Thanks for all your time and help! Really appreciate that!

dmuhamedagic commented:
Welcome!
